contrib/mesos/pkg/executor/registry.go

```diff
@@ -0,0 +1,305 @@
+	}
+	go r.b.run()
+	return r
+}
+
+func (r registryImpl) watch() <-chan *RegisteredPod {
+	x, err := r.b.listen()
+	if err != nil {
+		log.Error(err)
+		return nil
+	}
+	return x.in
+}
+
+func taskIDFor(pod *api.Pod) (taskID string, err error) {
+	taskID = pod.Annotations[meta.TaskIdKey]
+	if taskID == "" {
+		err = errUnknownTask
+	}
+	return
+}
+
+func (r registryImpl) shutdown() {
+	//TODO(jdef) flesh this out
+	r.m.Lock()
+	defer r.m.Unlock()
+	r.boundTasks = map[string]*api.Pod{}
+}
+
+func (r registryImpl) empty() bool {
+	r.m.RLock()
+	defer r.m.RUnlock()
+	return len(r.boundTasks) == 0
+}
+
+func (r registryImpl) pod(taskID string) *api.Pod {
+	r.m.RLock()
+	defer r.m.RUnlock()
+	return r.boundTasks[taskID]
+}
+
+func (r registryImpl) Remove(taskID string) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+	pod, ok := r.boundTasks[taskID]
+	if !ok {
+		return errUnknownTask
+	}
+
+	delete(r.boundTasks, taskID)
+
+	r.updates <- &RegisteredPod{
+		Pod:    pod,
+		taskID: taskID,
+		event:  RegistrationEventDeleted,
+	}
+
+	log.V(1).Infof("unbound task %v from pod %v/%v", taskID, pod.Namespace, pod.Name)
+	return nil
+}
+
+func (r registryImpl) Update(pod *api.Pod) (*RegisteredPod, error) {
+	// Don't do anything for pods without task anotation which means:
+	// - "pre-scheduled" pods which have a NodeName set to this node without being scheduled already.
+	// - static/mirror pods: they'll never have a TaskID annotation, and we don't expect them to ever change.
+	// - all other pods that haven't passed through the launch-task-binding phase, which would set annotations.
+	taskID, err := taskIDFor(pod)
+	if err != nil {
+		// There also could be a race between the overall launch-task process and this update, but here we
+		// will never be able to process such a stale update because the "update pod" that we're receiving
+		// in this func won't yet have a task ID annotation. It follows that we can safely drop such a stale
+		// update on the floor because we'll get another update later that, in addition to the changes that
+		// we're dropping now, will also include the changes from the binding process.
+		log.V(5).Infof("ignoring pod update for %s/%s because %s annotation is missing", pod.Namespace, pod.Name, meta.TaskIdKey)
+		return nil, err
+	}
+
+	r.m.Lock()
+	defer r.m.Unlock()
+	oldPod, ok := r.boundTasks[taskID]
+	if !ok {
+		return nil, errUnknownTask
+	}
+
+	registeredPod := &RegisteredPod{
+		Pod:    pod,
+		taskID: taskID,
+		event:  RegistrationEventUpdated,
+	}
+
+	// TODO(jdef) would be nice to only execute this logic based on the presence of
+	// some particular annotation.
+	//   - preserve the original container port spec since the k8sm scheduler
+	//   has likely changed it.
+	containers := oldPod.Spec.Containers
+	ctPorts := make(map[string][]api.ContainerPort, len(containers))
+	for i := range containers {
+		ctPorts[containers[i].Name] = containers[i].Ports
+	}
+	containers = pod.Spec.Containers
```

Comment:
this mutates the `pod` argument, which is bad practice. we should mutate a copy and store than copy in registeredPod. podsource can then use the copy from the returned result and pass it down the pipe to the kubelet.